### PR TITLE
Add CBMC harness for QueueCreateMutexStatic

### PR DESCRIPTION
**Pull request by mmuesly**
Description
-----------

This PR adds a CBMC harness for QueueCreateMutexStatic, QueueGenericCreateStatic and QueueCreateCountingSemaphoreStatic.

See the added README.mds for made assumptions.

Depends on #774 and #697

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by mmuesly** [comment]
The last force push adds configuration checking for `configSUPPORT_DYNAMIC_ALLOCATION` set to 0 and 1 for the `QueueGenericCreateStatic` harness.

File: tools/cbmc/patches/0003-Replace-MiniListItem_t-with-ListItem_t.patch [hunk 1]
```diff
@@ -1,21 +1,18 @@
...
-next element. CBMC gets confused when it sees a list of ListItem_ts with
-a smaller MiniListItem_t at the end, so this commit makes the last
-element of lists a ListItem_t instead.
-
-This patch should be removed soon: It is for a CBMC issue being fixed now.
----
- freertos_kernel/include/list.h | 2 +-
- 1 file changed, 1 insertion(+), 1 deletion(-)
-
+diff --git a/freertos_kernel/include/FreeRTOS.h b/freertos_kernel/include/FreeRTOS.h
```
Comment (danielsn) [inline tools/cbmc/patches/0003-Replace-MiniListItem_t-with-ListItem_t.patch]: Given that this bug appears to have been fixed, do we need the patch at all?

File: tools/cbmc/patches/0003-Replace-MiniListItem_t-with-ListItem_t.patch [hunk 2]
```diff
@@ -26,7 +23,7 @@ index 43e0ffffb..f6c6b6636 100644
 +	ListItem_t xListEnd;
  	listSECOND_LIST_INTEGRITY_CHECK_VALUE				/*< Set to a known value if configUSE_LIST_DATA_INTEGRITY_CHECK_BYTES is set to 1. */
  } List_t;
- 
--- 
+
```
Comment (danielsn) [inline tools/cbmc/patches/0003-Replace-MiniListItem_t-with-ListItem_t.patch]: Whitespace changes

File: tools/cbmc/proofs/Queue/QueueCreateCountingSemaphoreStatic/QueueCreateCountingSemaphoreStatic_harness.c [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    UBaseType_t uxMaxCount;
+    UBaseType_t uxInitialCount;
+
+    //xStaticQueue is required to be not null
+    StaticQueue_t xStaticQueue;
+
+    //Checked invariant
+    __CPROVER_assume(uxMaxCount != 0);
+    __CPROVER_assume(uxInitialCount <= uxMaxCount);
+    xQueueCreateCountingSemaphoreStatic( uxMaxCount, uxInitialCount, &xStaticQueue );
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueCreateCountingSemaphoreStatic/QueueCreateCountingSemaphoreStatic_harness.c]: Can we check that the semaphore was successfully created?

File: tools/cbmc/proofs/Queue/QueueCreateCountingSemaphoreStatic/QueueCreateCountingSemaphoreStatic_harness.c [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+
+void harness(){
+    UBaseType_t uxMaxCount;
+    UBaseType_t uxInitialCount;
+
+    //xStaticQueue is required to be not null
+    StaticQueue_t xStaticQueue;
+
+    //Checked invariant
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueCreateCountingSemaphoreStatic/QueueCreateCountingSemaphoreStatic_harness.c]: nice!

File: tools/cbmc/proofs/Queue/QueueCreateMutexStatic/QueueCreateMutexStatic_harness.c [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+#include "cbmc.h"
+
+
+void harness(){
+    uint8_t ucQueueType;
+
+    //The mutex storage is assumed to be not null.
+    StaticQueue_t xStaticQueue;
+
+    xQueueCreateMutexStatic( ucQueueType, &xStaticQueue );
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueCreateMutexStatic/QueueCreateMutexStatic_harness.c]: Same as before.  Can we check that the mutex is valid to use

File: tools/cbmc/proofs/Queue/QueueGenericCreateStatic/QueueGenericCreateStatic_harness.c [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    uint8_t *ucQueueStorage = (uint8_t *) pvPortMalloc(sizeof(uint8_t));
+
+    //The queue storage space is explicitly assumed to be not null.
+    StaticQueue_t xStaticQueue;
+    uint8_t ucQueueType;
+
+    /* The QueueGenericReset method called during static queue creation
+       implicitly that there are no overflows in the computation
+       and the inputs are safe. There is no check for this in the code base. */
+    UBaseType_t upper_bound = portMAX_DELAY - sizeof(Queue_t);
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueGenericCreateStatic/QueueGenericCreateStatic_harness.c]: I asked this in a previous PR.  portMAX_DELAY is an unclear name for the macro. Can you create a descriptive alias  for it

File: tools/cbmc/proofs/Queue/QueueGenericCreateStatic/README.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+The harness proves memory safety of
+QueueGenericCreateStatic under the assumption made in the harness.
+
+The principal assumption is that (uxItemSize * uxQueueLength) + sizeof(Queue_t)
+does not overflow. Further, ucQueueStorage must only be null iff uxItemSize is null.
+In addition, the passed queue storage is assumed to be allocated to the right size.
+
+The configurations for configSUPPORT_DYNAMIC_ALLOCATION set to 0 and 1 are checked.
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueGenericCreateStatic/README.md]: Are those both possible values?

File: tools/cbmc/proofs/make_type_header_files.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
```
Comment (danielsn) [inline tools/cbmc/proofs/make_type_header_files.py]: This doesn't belong in this PR

**Comment by markrtuttle** [comment]
Closed in favor of https://github.com/aws/amazon-freertos/pull/1324.
